proof/invariant-abstract/KHeap_AI.thy

```diff
@@ -2620,4 +2659,195 @@ lemma update_sched_context_no_fail[wp]:
   apply (clarsimp simp: obj_at_def a_type_def)
   done
 
+lemma update_reply_no_fail[wp]:
+  "no_fail (\<lambda>s. \<exists>reply. kheap s reply_ptr = Some (Reply reply)) (update_reply reply_ptr f)"
+  apply (clarsimp simp: update_reply_def)
+  apply (wpsimp wp: get_object_wp)
+  apply (clarsimp simp: obj_at_def)
+  done
+
+lemma update_reply_wp:
+  "\<lbrace>\<lambda>s. \<forall>reply. ko_at (Reply reply) reply_ptr s
+                \<longrightarrow> Q (s\<lparr>kheap := (kheap s)(reply_ptr \<mapsto> Reply (f reply))\<rparr>)\<rbrace>
+   update_reply reply_ptr f
+   \<lbrace>\<lambda>_. Q\<rbrace>"
+  by (wpsimp simp: update_reply_def wp: set_object_wp get_object_wp)
+
+lemma update_reply_obj_at_trivial:
+  "\<lbrace>obj_at P t' and K (\<forall>reply. P (Reply reply) \<longrightarrow> P (Reply (f reply)))\<rbrace>
+   update_reply reply_ptr f
+   \<lbrace>\<lambda>_. obj_at P t'\<rbrace>"
+  by (wpsimp wp: update_reply_wp simp: obj_at_def)
+
+lemma update_reply_typ_at[wp]:
+  "update_reply ptr f \<lbrace>\<lambda>s. P (typ_at T p s)\<rbrace>"
+  by (wpsimp simp: update_reply_def get_object_def wp: set_object_typ_at)
+
+lemmas update_reply_typ_ats [wp] = abs_typ_at_lifts [OF update_reply_typ_at]
+
+lemma update_reply_cte_wp_at[wp]:
+  "update_reply ptr f \<lbrace>cte_wp_at P p\<rbrace>"
+  by (wpsimp simp: update_reply_def set_object_def cte_wp_at_cases get_object_def)
+
+crunch update_reply
+  for ex_nonz_cap_to[wp]: "ex_nonz_cap_to p"
+  (wp: ex_nonz_cap_to_pres)
+
+lemma update_reply_pred_tcb_at[wp]:
+  "update_reply ptr update \<lbrace>\<lambda>s. P (pred_tcb_at proj f t s)\<rbrace>"
+  by (wpsimp simp: update_reply_def set_object_def get_object_def pred_tcb_at_def obj_at_def)
+
+lemma update_reply_valid_replies[wp]:
+  "update_reply ptr f \<lbrace>valid_replies_pred P\<rbrace>"
```

Comment:
is this proof copied from somewhere else? it uses `hoare_lift_Pf2` instead of `wps`, `wp` over `wpsimp`... looks ancient

Reply:
Yeah, almost all of the `update_reply` lemmas were copied from their equivalent `update_sched_context` versions, in an attempt to minimise the need to think. I'll go through your suggestions and try to apply them, it's always nice to at least not make things worse.